### DISCUSSION TITLE
Looking for help managing rollups in my org!

### DISCUSSION THREAD

**Discussion by GithubUser1892** [jamessimone/apex-rollup#430] [Q&A]
Hello,

I've recently started using Apex Rollup and would like some guidance on the best way to maintain these in my org. 

I have set up two scheduled flows, one at 1 AM and another at 4:30 AM. I made sure not to choose an object on the scheduled flows and am using the Full Recalc CMDT-driven Invocable action and a text template with the list of rollups.

I need some rollups to be done before the second set are rolled up, which is why I set them to run at different times. 

The issue I'm running into is: 

Failed to process batch for class 'RollupFullBatchRecalculator' for job id '707Mm000003pKfo'

caused by: System.LimitException: Apex heap size too large: 12230504

External entry point
Class.RollupLogger.log: line 44, column 1
Class.RollupLogger.CombinedLogger.log: line 186, column 1
Class.RollupAsyncProcessor.execute: line 231, column 1

I looked at similar posts and lowered the Batch Chunk Size to 200. There were 21,163 total batches and 6,964 batches processed.

The idea is that the first set of rollups would be finished before the second set ran, but since the first set started at 1 AM, I had to abort it 11 hours later as it was still running. The two scheduled flows would be ideally ran every night.

Thanks for your help!

**Comment by jamessimone** [comment]
Hi there - thanks for reaching out. I guess my first question would be "why do you need to do full recalcs every day?"

The intent of the full recalculation is to get data prepped in the org so that flow/Apex can then be responsible for maintaining the state of rollups in a much, much less process-intensive way. While I have tried to support the possibility of performing full recalculations across a broad swathe of different functionality, the reality is that for orgs with LDV _and_ many rollups (which it sounds like you may have both, given the heap size issue with a batch size of 500), it simply takes time for batch processes to run through all of the chunks in general.

As well - I would recommend disabling logging in production. It is the single largest use of resources when it comes to Apex Rollup, and especially for full recalcs, it's simply not an efficient usage of your org's limits. I would guess that with logging disabled, you'll see things run much more smoothly, and almost certainly would be able to re-up the Batch Chunk Size as well.

Lastly - are you using the latest version of Apex Rollup? Knowing which version you're on will help immensely; I've made some major improvements as of late to the full recalculation code paths, and if you're on an earlier version, upgrading will certainly help.

**Reply by GithubUser1892** [reply]
I am on the newest version of Apex Rollup! 

When we've tried to run them from a trigger, we would receive a lot of other errors from processes triggering the rollups and causing errors (Apex CPU time limit exceeded, too many SOQL queries, and NullPointerException) but would be open to trying it again if that's the recommended approach.

It doesn't look like we have logging enabled, but let me know if any of these settings are wrong.
![image](https://user-images.githubusercontent.com/130089442/230461007-206cbede-346d-40f3-8ca5-be6f2fbd7424.png)

**Reply by jamessimone** [reply]
Interesting - that timeout may simply be where in the code time is running out. That's unusual though.

Apex Rollup uses an absolute minimum of queries in any synchronous context, so I'm surprised to hear that you're running into limit exception issues. Many times I am able to help users of the package in the event that they're able to upload or send logs to me of their rollup runs. In general, the sort of things I tend to be looking out for are other long-running processes also part of the org, recursive automations, etc ...

**Reply by GithubUser1892** [reply]
In my scenario, what would the optimal setup for Apex Rollup be (LDV and many rollups)? Or how can I help you best troubleshoot the rollup runs?

Should I try to use the trigger or flow? Is one more efficient than the other? 

And how would I run the full recalculation, at least in the beginning?

**Reply by jamessimone** [reply]
I know many people who, even as flow enthusiasts, use the Apex one liner (`Rollup.runFromTrigger();`) in their trigger or trigger handler classes because it's easier than having to set up two record triggered flows minimum (create & update + delete). You should be able to browse to the `Recalculate Rollups` tab to do a one-off full recalculation, after which you would enable the trigger/flow version of Apex Rollup to keep maintaining fields.

As for troubleshooting, the easiest way is to have your Salesforce developer console open, with the Apex debug trace level set to `FINE`, and try running something like the existing scheduled flow that you're using. Once you get one `Batch Apex` log generated, that one log should have all of the relevant information in it (even with debug logging in Apex Rollup disabled). At the `FINE` logging level, debug logs show just how many times code is being called and how CPU time is being spent, which tends to paint a better picture than any description of the org. Let me know if that all makes sense - just know that I am committed to supporting your usage of Apex Rollup, and that I will work with you to address any issues we may uncover in the process.

**Reply by GithubUser1892** [reply]
Thank you so much for your support!! 

Another concern we had was with the Usage-based Entitlements - Maximum Flow Interviews Without UI per Month. We've used around 43,224,716 inverviews in the past 4 days and was wondering if this is something that is intended or taken into consideration, or if these are even related to Apex Rollup.

This is the log I got when I ran the scheduled flow, but I don't seem to be seeing any Batch Apex logs in my developer console, but I see that batches are being processed in the apex jobs tab. Any guidance here would be great.
[apex-07LMm000000AEGDMA4.log](https://github.com/jamessimone/apex-rollup/files/11173785/apex-07LMm000000AEGDMA4.log)

**Reply by GithubUser1892** [reply]
Found one: 
[apex-07LMm000000AEGDMA4.log](https://github.com/jamessimone/apex-rollup/files/11173840/apex-07LMm000000AEGDMA4.log)

**Reply by jamessimone** [reply]
I will take a look at that log. Scheduled Flow may run as the Automated Process User, which requires setting up another Trace from the Debug section in setup (you also have to uncheck the box for `Show only my logs` in the Debug menu in the Developer Console). There is an important note about scheduled Flow in the Apex Rollup readme which it may be pertinent to review here:

> In order to prevent blowing through the Flow Interview limit for each day, it's important to note that the use of Apex Rollup with a specific SObject in the scheduled flow's start node will run a flow interview for _every_ record retrieved. However, if the scheduled flow is run without a specific SObject having been selected in the start node, the process is bulkified successfully and you only consume a single flow interview per batch of records.

**Reply by GithubUser1892** [reply]
I made sure that there was not a specific SObject selected:
<img width="494" alt="Screenshot 2023-04-06 132439" src="https://user-images.githubusercontent.com/130089442/230475616-75d23667-9d03-4c35-88e1-278917741a64.png">

**Reply by jamessimone** [reply]
In that case, only a single flow interview should be getting used for the scheduled flow. however, each batch chunk from the batch jobs that are created will also consume a single flow interview (if you are using Flow to start up Apex Rollup). I will be able to take a look through the logs you've provided a bit later this afternoon and hopefully will be able to shed some additional light on what's going on.

**Reply by jamessimone** [reply]
It looks like both the attached log are only from the scheduled flow starting up, as opposed to anything that happens after that. I noticed though that there is some mention to Process Builder in the log - that would definitely be something I'd recommend looking into since Process Builder is not a bulkified automation tool and could really be eating into your overall CPU time limits!

**Reply by GithubUser1892** [reply]
Apologies, not sure why it sent the same log twice. 
[apex-07LMm000000AERaMAO.log](https://github.com/jamessimone/apex-rollup/files/11174120/apex-07LMm000000AERaMAO.log)

I'm not sure why it would mention process builder in the log. There are no process builder automations in the org. When using the Full Recalc CMDT-driven Invocable action, is there a preferred method of comma-separated list? I wonder if it's the text template?

**Reply by jamessimone** [reply]
Ah interesting, I wonder why the log was denoting the existence of Process Builder? That's a new one to me, and I've been looking through a lot of these logs recently. Apologies for the red herring there. 

I don't see anything particularly note-worthy as far as how timeouts could be occurring, though - the entire batch is processed in under 3 seconds, which means there's 57 seconds of CPU time to spare. DML and all that look to be in very safe spots as well as far as limits are concerned. If you are able to reproduce one of the errors you were talking about and get a log generated for one of those, that might yield further insight?

**Reply by GithubUser1892** [reply]
Good morning,

Do you know if there's a way to download a log based on a job id? It looks like all the errors happened at night, so I would have to wait until they started again. The scheduled job I ran yesterday that you requested is still running with a Status Detail of:

First error: common.exception.SfdcSqlException: An I/O error occurred while sending to the backend. [R7rHoFZ5Qm2GffmEcAMSUw]; [conn=STANDARD:1:1:85986:1]

With 7,143 Total Batches, 3,996 Batches Processed, and 1 Failures.

Not sure if related to this exact scheduled job, but I received a new exception as well:

Failed to process batch for class 'RollupFullBatchRecalculator' for job id '707Mm000003qTaf'

caused by: System.LimitException: Batchable instance is too big: RollupParentResetProcessor.QueueableResetProcessor

External entry point
Class.RollupParentResetProcessor.runCalc: line 56, column 1
Class.RollupFullRecalcProcessor.finish: line 77, column 1
Class.RollupAsyncProcessor.finish: line 253, column 1

**Reply by GithubUser1892** [reply]
I was able to get some logs!
[apex-07LMm000000AGiAMAW.log](https://github.com/jamessimone/apex-rollup/files/11179532/apex-07LMm000000AGiAMAW.log)
[apex-07LMm000000AGlNMAW.log](https://github.com/jamessimone/apex-rollup/files/11179533/apex-07LMm000000AGlNMAW.log)

**Reply by jamessimone** [reply]
Amazing. Let me take a look through the logs and see what else I can see.

**Reply by jamessimone** [reply]
One thing I can recommend (to help deal with the issue with the ParentResetProcessor), is to set the `Should Skip Resetting Parent Fields` checkbox on the Org Default rollup control CMDT to true. That will prevent the "batchable instance is too big" error.

**Reply by GithubUser1892** [reply]
I've made that update! Any luck on the Apex heap size too large exception? 

Or do you think if I manually run the recalculations through the Recalculate Rollups tab that should work for the one-time setup?

**Reply by jamessimone** [reply]
That change will remove the Apex heap size being too large exception entirely because the `RollupParentResetProcessor` won't be in use - I haven't had the chance yet to take a look through the latest logs you've provided, but I am hopeful to be able to do so later today or this weekend.

**Reply by jamessimone** [reply]
@GithubUser1892 in the event that updating the flag doesn't completely solve the heap size issue, I've identified some additional heap savings that are (presumably) related to the error you were getting, which I intend to ship in #431 (hoping to release that today)

**Reply by GithubUser1892** [reply]
That's great! I'll update to the newest release as well!

Another question I had, is it okay to include rollups from different objects in the same Full Recalc CMDT-driven Invocable?

**Reply by jamessimone** [reply]
Yes, it's perfectly okay. Depending on the number of children records for each full recalc you're using, if Apex Rollup sees that the total number of children is over the `Max Lookup Rows Prior To Batching` number on the org default Rollup Control record, a batch job will be spawned per child object type. These batch jobs are processed serially - that is, one at a time - so that's just something to be aware of if you're using the Apex Jobs page in setup to track their progress.
